File: src/station/harvest.py
"""The Station biomass/food loop: biosphere grain → crew ``food_store`` (P6.6).

Step 6's assembly — built on the Step-3 :mod:`station.greenhouse` (the only assembly
where a live plant already shares the cabin air, so it is where biomass can flow into
food). The seam adds **one** station-owned flow, :class:`station.flows.Harvest`
(``storage_c → food_store``, donor-controlled), to the cabin / fast registry: the
biosphere fixes cabin CO₂ into grain (``storage_c``) and the harvest moves that grain
into the crew's finite ``food_store``, so the store — open-loop and monotonically
depleting in standalone Crew / the greenhouse — becomes **regenerative** up to the
harvest it captures. It is the **CARBON twin of Step 4's** ``WaterRecovery`` (WATER),
one trophic level.

**The plan splits this into two seams (advisor-flagged, landed as separate commits).**
Seam 1 is the ``Harvest`` flow (``with_harvest``). Seam 2 (``close_feces``) re-points
``CrewRespiration``'s fecal carbon into the biosphere's ``LITTER_CARBON`` pool, closing
the trophic CARBON ring (feces → litter → microbes → CO₂). They were landed separately
because crew-scale feces (~363 mol C / 7 days) dumped into the seedling-scale litter
pool **dominates** the litter dynamics (~3400× mismatch), and that had to be
understood on its own. The measured regime is **active microbial** (not throttled):
litter → ~342 mol, microbial biomass → ~20 mol, so ``Δlitter ≈ feces`` does *not*
hold (microbes consume ~21 mol) — the seam-2 gate is closure + no-shadow-sink +
litter-grows-materially, not a three-way identity (calibration of the magnitude
mismatch deferred to Step 9).

**Why the reproductive plant.** ``storage_c`` fills only after anthesis (``FO > 0``
needs ``DVS > 1``), and the default seedling sits at ``DVS < 1`` with ``storage_c0 = 0``
— a zero harvest source. :attr:`station.scenario.HarvestScenario.thermal_time0` starts
the biosphere phenology accumulator **past** anthesis (``tsum_anthesis = 1100`` °C·day),
so grain is actively *filling* while harvest drains it — a genuinely regenerative
source. This is injected at :class:`State` construction here (the station owns the
greenhouse ``State``'s ``thermal_time`` aux), so ``SeasonScenario`` is untouched (a
``thermal_time0`` field there would be a domain change, forbidden by the Phase-6 exit
criterion).

**The exact with/without-harvest identity (the "it bit" payload).** The ``Allocation``
grain-fill leg (``FO·DMI``) is independent of ``storage_c``'s own level, and no other
biosphere flow reads ``storage_c`` (only ``annual_reset``, which does not fire in a
≤7-day run), so harvest is its **only** new sink and does not perturb the plant's carbon
budget. Grain fill is therefore *identical* with and without harvest, so the two-way
identity ``Δfood_store = cumulative harvest = Δstorage_c`` holds to floating point (the
``with_harvest`` baseline arm drops the ``Harvest`` flow: ``storage_c`` then accumulates
all fill and ``food_store`` depletes at the full crew rate). ``CrewRespiration`` is
**forced** (reads the intake schedule, not ``food_store``), so the regenerated (higher)
``food_store`` in the harvest arm does not perturb the crew CO₂ draw → the cabin gas
trajectory is identical too.

**The two-rate driver + Euler-only.** Runs under the shared
:func:`station.driver.run_master_day` (the greenhouse rhythm: biosphere slow once/day —
refilling grain — then cabin fast ×``steps_per_day`` — draining it), so the day-boundary
snapshot is the intra-day *minimum* ``storage_c`` (post-drain), which the harvest rate
is sized to keep positive (a regenerative source, not a static reservoir emptied). The
greenhouse biosphere is **Euler-locked by its freeze**, so this is an Euler-only run (no
RK4 cross-check — the ``WaterRecovery`` "state-dependent breaks RK4≡Euler" signal does
not apply on a biosphere-coupled build).

Pure stdlib only in the spine; the harvest rate loads via ``station.loader``. Reuses
``build_greenhouse`` wholesale (zero greenhouse change) — this module only adds the one
flow + the phenology injection.
"""

from domains.biosphere.stocks import (
    LITTER_CARBON,
    STORAGE_C,
    THERMAL_TIME,
    VERNALIZATION_DAYS,
)
from domains.crew.flows import CrewParams
from domains.crew.stocks import FECAL_WASTE, FOOD_STORE
from domains.eclss.flows import EclssParams
from simcore.environment import SourceResolver
from simcore.events import Event
from simcore.ids import FlowId
from simcore.integrator import EulerIntegrator
from simcore.registry import Registry
from simcore.state import State
from station.driver import run_master_day
from station.flows import Harvest, HarvestParams
from station.greenhouse import (
    build_greenhouse,
    greenhouse_bio_resolver,
    greenhouse_cabin_resolver,
)
from station.scenario import HARVEST_SCENARIO, HarvestScenario

# The station-owned harvest flow id (the biosphere-grain → crew-food seam). Distinct
# from every cabin flow id (``station.crew_respiration`` etc.) and every biosphere flow
# id — the flow-registry disjointness assert in ``build_harvest`` enforces it.
HARVEST: FlowId = FlowId("station.harvest")


def build_harvest(
    crew_params: CrewParams,
    eclss_params: EclssParams,
    harvest_params: HarvestParams,
    scenario: HarvestScenario = HARVEST_SCENARIO,
    *,
    with_harvest: bool = True,
    close_feces: bool = True,
) -> tuple[State, Registry, Registry]:
    """Assemble the harvest greenhouse: ``(state, bio_reg, cabin_reg)``.

    Reuses :func:`station.greenhouse.build_greenhouse` (the sealed biosphere ↔ cabin gas
    loop — zero greenhouse-behavior change; the greenhouse gains only an optional
    ``fecal_waste_target`` param, default-preserving) and layers on the Step-6
    additions:

      1. the biosphere ``thermal_time`` aux is started at ``scenario.thermal_time0``
         (past anthesis ⇒ a grain-filling plant), overriding the greenhouse's ``0.0`` —
         the station owns the ``State``'s aux dict, so ``SeasonScenario`` stays
         untouched;
      2. the :class:`station.flows.Harvest` flow (``storage_c → food_store``) is
         appended to the cabin / fast registry (``with_harvest=True`` — seam 1); and
      3. ``close_feces=True`` (seam 2) re-points ``CrewRespiration``'s fecal carbon into
         the biosphere's ``LITTER_CARBON`` pool (via ``build_greenhouse``'s
         ``fecal_waste_target``), closing the trophic CARBON ring (feces → litter →
         microbes → CO₂) and **omitting** the orphaned ``FECAL_WASTE`` boundary sink (no
         shadow sink). ``close_feces=False`` keeps feces on the open ``FECAL_WASTE``
         sink — the "it bit" baseline the seam-2 gate compares against.

    ``with_harvest=False`` is the seam-1 baseline (no ``Harvest`` flow — ``storage_c``
    accumulates all fill, ``food_store`` depletes at the full crew rate).

    **Seam-2 regime (measured, not assumed — advisor-flagged).** Crew-scale feces (~363
    mol C / 7 days) dumped into the seedling-scale ``LITTER_CARBON`` (starts 0) makes
    ``MicrobialRespiration`` **active** (litter → ~342 mol, microbial biomass → ~20
    mol), not throttled — so ``Δlitter ≈ feces`` does **not** hold (microbes consume ~21
    mol of the routed carbon). The gate is therefore per-quantity closure +
    ``FECAL_WASTE`` sink absent + litter grows materially (with vs without the re-point)
    + ``rationed == 0`` (the once-daily microbial O₂ draw is refilled by ``O2Makeup``) +
    ``events == ()`` —
    **not** a three-way identity. This ~3400× crew-vs-plant magnitude mismatch is the
    documented scope boundary (calibration deferred to Step 9).

    The two flow registries (biosphere-slow vs cabin-fast) are asserted to have
    **disjoint flow-id sets** over the one shared stock dict: a duplicate ``FlowId``
    across the two registries would be a silent cross-domain wiring bug (each
    ``Registry`` already rejects a duplicate *within* itself, but not across the pair
    the driver steps together).
    """
    state, bio_reg, cabin_reg = build_greenhouse(
        crew_params,
        eclss_params,
        scenario.greenhouse,
        with_plants=True,
        fecal_waste_target=LITTER_CARBON if close_feces else FECAL_WASTE,
    )

    # (1) Start the biosphere phenology past anthesis (a grain-filling plant). The
    # station owns the greenhouse State's aux dict, so this is a station-level injection
    # — no SeasonScenario / domain change.
    state = State(
        n=state.n,
        stocks=state.stocks,
        rng_seed=state.rng_seed,
        # thermal_time0 starts the crop PAST anthesis (DVS > 1), where the
        # vernalization factor is fixed at 1 — so a zero cold accumulator here is
        # inert by construction, not an omission.
        aux={THERMAL_TIME: scenario.thermal_time0, VERNALIZATION_DAYS: 0.0},
    )

    # (2) Append the Harvest flow to the cabin / fast registry (drained across the day's
    # sub-steps; the biosphere refills grain in its once-daily slow step).
    cabin_flows = list(cabin_reg.flows)
    if with_harvest:
        cabin_flows.append(
            Harvest(
                HARVEST,
                0,
                storage_c=STORAGE_C,
                food_store=FOOD_STORE,
                params=harvest_params,
            )
        )
    cabin_reg = Registry(cabin_flows, state.stocks)

    _assert_flow_ids_disjoint(bio_reg, cabin_reg)
    return state, bio_reg, cabin_reg


def _assert_flow_ids_disjoint(bio_reg: Registry, cabin_reg: Registry) -> None:
    """Guard: the biosphere-slow and cabin-fast registries share no ``FlowId``.

    The driver steps both registries over one shared stock dict; a flow id present in
    both would be a silent cross-domain wiring bug (a ``Registry`` rejects duplicates
    only *within* itself). This is the flow-id analogue of the greenhouse's stock-id
    disjointness check.
    """
    bio_ids = {flow.id for flow in bio_reg.flows}
    cabin_ids = {flow.id for flow in cabin_reg.flows}
    overlap = bio_ids & cabin_ids
    if overlap:
        raise ValueError(
            "harvest flow-id collision between the biosphere and the cabin registries: "
            f"{sorted(overlap)} (the two flow sets the driver steps together must be "
            "disjoint)"
        )


def harvest_bio_resolver(
    weather: list[dict[str, float | str]],
    scenario: HarvestScenario = HARVEST_SCENARIO,
) -> SourceResolver:
    """The biosphere forcing resolver — the greenhouse's, over the embedded scenario.

    Delegates to :func:`station.greenhouse.greenhouse_bio_resolver` (the frozen
    ``weather_resolver`` + the default sealed shared map, ``CO2_POOL_VAR → CARBON_POOL``
    = the cabin air). Step 6 changes no biosphere forcing — only the ``thermal_time``
    initial condition (set in :func:`build_harvest`) and the added harvest sink.
    """
    return greenhouse_bio_resolver(weather, scenario.greenhouse)


def harvest_cabin_resolver(
    scenario: HarvestScenario = HARVEST_SCENARIO,
) -> SourceResolver:
    """The cabin forcing resolver — the greenhouse's two constant crew intake rates.

    Delegates to :func:`station.greenhouse.greenhouse_cabin_resolver`. The ``Harvest``
    flow is **donor-controlled** (reads ``storage_c``, needs no forcing var), so no new
    forcing is added.
    """
    return greenhouse_cabin_resolver(scenario.greenhouse)


def run_harvest(
    bio_integrator: EulerIntegrator,
    cabin_integrator: EulerIntegrator,
    state: State,
    bio_resolver: SourceResolver,
    cabin_resolver: SourceResolver,
    scenario: HarvestScenario = HARVEST_SCENARIO,
) -> tuple[list[State], int, tuple[Event, ...]]:
    """The two-rate master-step driver: one day per step (biosphere-slow / cabin-fast).

    A thin wrapper over :func:`station.driver.run_master_day` with the biosphere as the
    **slow** domain (once/day, advancing ``thermal_time`` — so grain refills — and
    ``n``) and the cabin (now carrying the ``Harvest`` flow) as the **fast** domain
    (``substep`` ×``steps_per_day``, ``n`` kept — so grain drains across the day).
    ``states`` holds one entry per day boundary (the golden pins the final one);
    ``total_rationed`` sums the Euler-backstop firings (validation asserts ``== 0``);
    ``events`` are extinction events (empty — the crew has no POPULATION stock, and the
    plant is well past anthesis but short of maturity over the horizon). Euler-only (the
    biosphere is Euler-locked by its freeze), so the integrators are
    :class:`EulerIntegrator`.

    Timing comes from the embedded greenhouse scenario (``days`` / ``steps_per_day`` /
    ``cabin_dt`` / ``bio_dt``), so the driver's ``fast_dt·steps_per_day == 86400`` (one
    day) invariant holds by construction.
    """
    gh = scenario.greenhouse
    return run_master_day(
        bio_integrator,
        cabin_integrator,
        state,
        bio_resolver,
        cabin_resolver,
        days=gh.days,
        steps_per_day=gh.steps_per_day,
        slow_dt=gh.bio_dt,
        fast_dt=gh.cabin_dt,
    )

File: rust/crates/station/src/harvest.rs
//! The biomass/food loop — the port of `station.harvest` (P6.6 / P7.5).
//!
//! Built on the Step-3 greenhouse. Adds **one** station-owned flow, [`Harvest`]
//! (`storage_c → food_store`, donor-controlled), to the cabin / fast registry — the CARBON
//! twin of Step-4 `WaterRecovery`, making the crew's finite `food_store` **regenerative**.
//! The reproductive plant precondition is met by injecting the biosphere `thermal_time` aux
//! **past anthesis** at `State` construction (a station-level injection — `SeasonScenario`
//! is untouched). Seam 2 (`close_feces`) re-points `CrewRespiration`'s fecal carbon into
//! `LITTER_CARBON`, closing the trophic CARBON ring. Two-rate, Euler-only. Tier-2 (FvCB).

use std::collections::BTreeMap;

use domains::biosphere::science;
use domains::biosphere::stocks::{
    CARBON_POOL, LITTER_CARBON, O2_POOL, ROOTED_DEPTH, SOIL_WATER, STORAGE_C, SUBSOIL_WATER,
    THERMAL_TIME,
};
use domains::crew::{CrewParams, FECAL_WASTE, FOOD_STORE};
use domains::eclss::EclssParams;
use simcore::environment::SourceResolver;
use simcore::error::SimError;
use simcore::events::Event;
use simcore::integrator::EulerIntegrator;
use simcore::registry::Registry;
use simcore::state::State;

use crate::cabin::build_cabin_flows;
use crate::driver::run_master_day;
use crate::flows::{Harvest, HarvestParams, HARVEST};
use crate::greenhouse::{build_greenhouse, greenhouse_bio_resolver, greenhouse_cabin_resolver};
use crate::scenario::HarvestScenario;

/// Assemble the harvest greenhouse: `(state, bio_reg, cabin_reg)`.
///
/// Reuses [`build_greenhouse`] (the sealed biosphere ↔ cabin gas loop), then: (1) starts the
/// biosphere `thermal_time` aux at `scenario.thermal_time0` (past anthesis ⇒ grain-filling);
/// (2) appends the [`Harvest`] flow to the cabin / fast registry (`with_harvest`); and (3)
/// `close_feces` re-points fecal carbon into `LITTER_CARBON` (omitting the `FECAL_WASTE`
/// sink). The bio/cabin **flow-id** sets are asserted disjoint.
pub fn build_harvest(
    crew: &CrewParams,
    eclss: &EclssParams,
    harvest: &HarvestParams,
    scenario: &HarvestScenario,
    with_harvest: bool,
    close_feces: bool,
) -> Result<(State, Registry, Registry), SimError> {
    let fecal_target = if close_feces {
        LITTER_CARBON
    } else {
        FECAL_WASTE
    };
    let (gh_state, bio_reg, _gh_cabin_reg) =
        build_greenhouse(crew, eclss, &scenario.greenhouse, true, fecal_target)?;

    // (1) Start the biosphere phenology past anthesis (a grain-filling plant) — a
    // station-level aux injection over the greenhouse State's stocks.
    // ⚠ THE INJECTED DEPTH DRAGS THE WATER STORES WITH IT (2026-08-12). Injecting a
    // 1.3 m root system while the stocks still hold the SOWING zone's water is incoherent
    // once the store is geometric: the biosphere seeds `soil_water0` for a 0.15 m zone,
    // and that inside a 1.3 m zone is `FTSW = 0.115` — a grain-filling crop declared at a
    // third of its growth threshold on day 0. Re-derived here from the same [F] Eqns
    // 14.26-14.28 the scenario defaults use, exactly as Python does.
    let bio = &scenario.greenhouse.bio;
    let atsw = science::captured_water(
        scenario.rooted_depth0,
        bio.soil_extractable_water,
        bio.ground_area,
    ) * bio.soil_moisture_index;
    let wstorg = science::captured_water(
        bio.soil_depth - scenario.rooted_depth0,
        bio.soil_extractable_water,
        bio.ground_area,
    ) * bio.soil_moisture_index;
    let mut stocks = gh_state.stocks.clone();
    let sw = stocks[SOIL_WATER].with_amount(atsw)?;
    stocks.insert(SOIL_WATER.to_string(), sw);
    let sub = stocks[SUBSOIL_WATER].with_amount(wstorg)?;
    stocks.insert(SUBSOIL_WATER.to_string(), sub);
    let state = State::new(
        gh_state.n,
        stocks,
        gh_state.rng_seed,
        BTreeMap::from([
            (THERMAL_TIME.to_string(), scenario.thermal_time0),
            // Goes with thermal_time0: a crop started past anthesis has finished rooting.
            (ROOTED_DEPTH.to_string(), scenario.rooted_depth0),
        ]),
    )?;

    // (2) Rebuild the cabin flows (the Rust Registry does not lend out owned flows) and
    // append Harvest — mirroring Python's `list(cabin_reg.flows) + [Harvest(...)]`.
    let mut cabin_flows = build_cabin_flows(crew, eclss, CARBON_POOL, O2_POOL, fecal_target);
    if with_harvest {
        cabin_flows.push(Box::new(Harvest::new(
            HARVEST.to_string(),
            STORAGE_C.to_string(),
            FOOD_STORE.to_string(),
            *harvest,
        )));
    }
    let cabin_reg = Registry::flows_only(cabin_flows, &state.stocks)?;

    assert_flow_ids_disjoint(&bio_reg, &cabin_reg)?;
    Ok((state, bio_reg, cabin_reg))
}

/// Guard: the biosphere-slow and cabin-fast registries share no `FlowId`.
fn assert_flow_ids_disjoint(bio_reg: &Registry, cabin_reg: &Registry) -> Result<(), SimError> {
    let bio_ids: std::collections::BTreeSet<&str> =
        bio_reg.flows().iter().map(|f| f.id()).collect();
    for flow in cabin_reg.flows() {
        if bio_ids.contains(flow.id()) {
            return Err(SimError::Validation(format!(
                "harvest flow-id collision between the biosphere and the cabin registries: \
                 {:?} (the two flow sets the driver steps together must be disjoint)",
                flow.id()
            )));
        }
    }
    Ok(())
}

/// The biosphere forcing resolver — the greenhouse's, over the embedded scenario.
pub fn harvest_bio_resolver(scenario: &HarvestScenario) -> Result<SourceResolver, SimError> {
    greenhouse_bio_resolver(&scenario.greenhouse)
}

/// The cabin forcing resolver — the greenhouse's two constant crew intake rates.
pub fn harvest_cabin_resolver(scenario: &HarvestScenario) -> Result<SourceResolver, SimError> {
    greenhouse_cabin_resolver(&scenario.greenhouse)
}

/// The two-rate driver: one day per step (biosphere-slow / cabin-fast).
pub fn run_harvest(
    bio_integrator: &EulerIntegrator,
    cabin_integrator: &EulerIntegrator,
    state: State,
    bio_resolver: &SourceResolver,
    cabin_resolver: &SourceResolver,
    scenario: &HarvestScenario,
) -> Result<(Vec<State>, u64, Vec<Event>), SimError> {
    let gh = &scenario.greenhouse;
    run_master_day(
        bio_integrator,
        cabin_integrator,
        state,
        bio_resolver,
        cabin_resolver,
        gh.days,
        gh.steps_per_day,
        gh.bio_steps_per_day,
        gh.bio_dt,
        gh.cabin_dt,
        None,
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::params as station_params;
    use crate::scenario::harvest_scenario;
    use domains::biosphere::science;
    use domains::params;

    /// The station's past-anthesis injection keeps DEPTH and WATER together — the place
    /// this gap actually bit.
    ///
    /// `build_harvest` overrides `rooted_depth0` to 1.3 m on top of a greenhouse built for
    /// the 0.15 m sowing zone. Before 2026-08-12 it inherited that zone's water: 19.5 kg
    /// inside a 169 kg capacity, `FTSW = 0.115`, grain 79 % low — a grain-filling crop
    /// declared at a third of its growth threshold on day 0. Both stores are now re-derived
    /// from the injected depth, and this asserts the resulting STATE rather than the code
    /// path, so a future refactor cannot quietly drop it.
    ///
    /// ⚠ This is the concrete case behind
    /// `scenario.rs`'s `every_scenarios_water_stores_are_geometric`: that census covers
    /// declared scenarios, and the harvest injection is a store rewritten at BUILD time,
    /// which no scenario-level census can see.
    /// Mirrors `tests/test_soil_layers.py::test_the_harvest_injection_keeps_depth_and_
    /// water_together`.
    #[test]
    fn the_harvest_injection_keeps_depth_and_water_together() {
        let crew = params::crew();
        let eclss = params::eclss();
        let harvest_params = station_params::harvest();
        let scenario = harvest_scenario();
        let (state, _bio, _cabin) =
            build_harvest(&crew, &eclss, &harvest_params, &scenario, true, true)
                .expect("build_harvest");
        let bio = &scenario.greenhouse.bio;
        let depth = state.aux[ROOTED_DEPTH];
        assert_eq!(depth, scenario.rooted_depth0);

        let capacity = science::captured_water(depth, bio.soil_extractable_water, bio.ground_area);
        let held = state.stocks[SOIL_WATER].amount;
        assert!(
            (held - capacity * bio.soil_moisture_index).abs() <= 1e-12 * capacity,
            "the injected crop's root zone holds {held}, not its own geometry {capacity}"
        );
        // ...which is to say it starts at its DECLARED FTSW, not at 0.115.
        let ftsw = held / capacity;
        assert!(
            (ftsw - bio.soil_moisture_index).abs() <= 1e-12,
            "the injected crop starts at FTSW {ftsw}, not the declared MAI {}",
            bio.soil_moisture_index
        );
        assert!(
            ftsw > 0.5,
            "the 0.115 regression is back: a grain-filling crop below its growth threshold"
        );
        // The below-root store follows the same re-derivation, or the profile stops being
        // a partition and the injection creates or destroys water.
        let below = state.stocks[SUBSOIL_WATER].amount;
        let want = science::captured_water(
            bio.soil_depth - depth,
            bio.soil_extractable_water,
            bio.ground_area,
        ) * bio.soil_moisture_index;
        assert!(
            (below - want).abs() <= 1e-12 * want,
            "the below-root store {below} is not (SOLDEP - DEPORT) geometry {want}"
        );
        // ⚠ NON-VACUITY: the injection must actually have MOVED the stores. If the
        // greenhouse's own `soil_water0` happened to equal the injected value, every
        // assertion above would hold on a build that had dropped the re-derivation.
        assert_ne!(
            bio.soil_water0, held,
            "the injected depth did not change the store, so this test proves nothing"
        );
    }
}
